plugins/imfile/imfile.c

```diff
@@ -543,6 +546,81 @@ wdmapDel(const int wd)
 
 #endif // #ifdef HAVE_INOTIFY_INIT
 
+/*
+ * Symbolic link support
+ */
+static struct hashtable *slink_target;     /* hashtable for symlink to target */
+static struct hashtable *target_slink;     /* hashtable for target to symlink */
+
+/*
+ * input:  file - either partial or full path
+ * input:  dir  - file's parent dir
+ * output: target - if not NULL, allocated target name in full path will be returned.
+ * return value: is_symlink: 1 - path is a symlink
+ *                           0 - path is not a symlink
+ *                          -1 - failed to get the info
+ *                          RS_RET_OUT_OF_MEMORY - out of memory
+ */
+static int
+isSymlink(char *file, char *dir, char **target)
+{
+	const char fullpath[MAXFNAME];
+	const char *ptr = fullpath;
+	struct stat fileInfo;
+	int is_symlink = 0;
+
+	if (file[0] == '/') {
+		ptr = file;
+	} else {
+		snprintf((char *)fullpath, MAXFNAME, "%s/%s", dir, file);
```

Comment:
```
fullpath[MAXFNAME-1] = '\0';
```
to ensure null termination

Reply:
@richm I had a discussion today on the need to do this sanity check. I also think it's the right thing to do, but I could not find any evidence we are right. Do you have evidence?

Reply:
@rgerhards I've always had to ensure proper null termination of strings when using snprintf.  I don't have any evidence, just many years of doing it this way.

Reply:
@richm I feared that answer... much like I knew and many do. But may look like it is no longer true. Should you run over some case where it does not work, let me know ;-)

Reply:
> may look like it is no longer true.

How so?  Was snprintf changed to always do null termination?  To my knowledge (man snprintf on Fedora 27) it was no.  Was there some change to the compiler to ensure `const char fullpath[MAXFNAME]` is initialized to all nulls?  again, not to my knowledge.

Reply:
@richm actually yes, snprintf is specified to do that in C99 (here a link to POSIX): http://pubs.opengroup.org/onlinepubs/007904875/functions/fprintf.html

It is interesting that the man page actually does not make a clear guarantee in that regard. I checked the Fedora 27 man pages, as well as Ubuntu and BSD, and indeed they do not specify it precisely - but they claim C99 compatibility. Web searches often bring up discussions where it is said this could be treated as granted.

Red Hat's Florian Weimer wrote a longer post about the problems with ```snprintf```: https://access.redhat.com/blogs/766093/posts/1976193 In it, he states
```
This code assumes that the snprintf implementation does not write an unterminated string to the destination buffer on error, which is quite reasonable as far as such assumptions go.  
```
Given his obviously in-depth resarch, I would guess that he would have noticed if there is a problem.

In conclusion, to me it looks like this is really something that we can rely on - but I am still puzzled. As I said, I was also thinking that there may be a problem on some platforms, but I cannot find hard evidence of where these problems really exist. Having pointers would be most useful.

Reply:
I should have also provided this quote from POSIX:

*The snprintf() function shall be equivalent to sprintf(), with the addition of the n argument which states the size of the buffer referred to by s. If n is zero, nothing shall be written and s may be a null pointer. Otherwise, output bytes beyond the n-1st shall be discarded instead of being written to the array, and a null byte is written at the end of the bytes actually written into the array.*

It's pretty much at the top of the document.

Reply:
Ok.  That seems to be the behavior on fedora 27 (gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)) and rhel7 (gcc (GCC) 4.8.5 20150623 (Red Hat 4.8.5-16)).  Trying various ways to break snprintf with a simple test program shows that it always null terminates the string and returns the correct result code.

Consider my objection withdrawn.